### PR TITLE
Functions on the registered runtime

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

**Comment by systay** [comment]
Unit tests for the RegisterAllocation additions still missing

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/BuildEnterpriseInterpretedExecutionPlan.scala [hunk 1]
```diff
@@ -56,7 +56,7 @@ object BuildEnterpriseInterpretedExecutionPlan extends Phase[EnterpriseRuntimeCo
     try {
       val (logicalPlan, pipelines) = rewritePlan(context, from.logicalPlan)
       val idMap = LogicalPlanIdentificationBuilder(logicalPlan)
-      val converters = new ExpressionConverters(CommunityExpressionConverter, EnterpriseExpressionConverters)
+      val converters = new ExpressionConverters(EnterpriseExpressionConverters, CommunityExpressionConverter)
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/BuildEnterpriseInterpretedExecutionPlan.scala:59]: Surprised this did not cause issues before.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisteredRewriter.scala [hunk 2]
```diff
@@ -186,6 +202,32 @@ class RegisteredRewriter(tokenContext: TokenContext) {
...
+      case (LongSlot(offset, _, typ, name), None) if typ == CTNode =>
+        NodePropertyExistsLate(offset, propKey, s"$name.$propKey")
+
+      case (LongSlot(offset, _, typ, name), Some(token)) if typ == CTRelationship =>
+        RelationshipPropertyExists(offset, token, s"$name.$propKey")
+
+      case (LongSlot(offset, _, typ, name), None) if typ == CTRelationship =>
+        RelationshipPropertyExistsLate(offset, propKey, s"$name.$propKey")
+
+      case _ => throw new CantCompileQueryException(s"Expressions with on object other then nodes and relationships are not yet supported")
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisteredRewriter.scala]: Perhaps remove 'with'

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/EnterprisePipeBuilder.scala [hunk 3]
```diff
@@ -99,34 +99,34 @@ class EnterprisePipeBuilder(fallback: PipeBuilder,
         ProduceResultRegisterPipe(source, runtimeColumns)(id)
 
       case Expand(_, IdName(from), dir, types, IdName(to), IdName(relName), ExpandAll) =>
-        val fromSlot = pipeline(from)
-        val relSlot = pipeline(relName)
-        val toSlot = pipeline(to)
-        ExpandAllRegisterPipe(source, fromSlot.offset, relSlot.offset, toSlot.offset, dir, LazyTypes(types), pipeline)(id)
+        val fromSlot = pipeline.getLongOffsetFor(from)
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/EnterprisePipeBuilder.scala:103]: Variables should be renamed back to 'offset' also then.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/PipelineInformation.scala [hunk 4]
```diff
@@ -20,15 +20,113 @@
...
 
+import scala.collection.mutable
 
 object PipelineInformation {
   def empty = new PipelineInformation(Map.empty, 0, 0)
 
   def apply(slots: Map[String, Slot], numberOfLongs: Int, numberOfReferences: Int) =
     new PipelineInformation(slots, numberOfLongs, numberOfReferences)
+
+  def toString(startFrom: LogicalPlan, m: Map[LogicalPlan, PipelineInformation]): String = {
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/PipelineInformation.scala:35]: I'm not seeing any calls to this new PipelineInformation.toString(....) method in the PR. Is this part of some other work?

**Review by craigtaverner — CHANGES_REQUESTED** [review]
Surely some of the cucumber tests that previously failed for slotted runtime would start passing now? I did not see changes to the blacklisting here.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/PipelineInformation.scala (hunk 4 above)
Comment (systay) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/PipelineInformation.scala:35]: You are right. It helped me debug a problem, and I thought it was helpful enough to stick around, but I'm not sure how to expose it to the public yet.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/BuildEnterpriseInterpretedExecutionPlan.scala (hunk 1 above)
Comment (systay) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/BuildEnterpriseInterpretedExecutionPlan.scala:59]: Before the two converters never took on the same expressions, so the order was not important. With these new specialisations, that is no longer true.

**Review by craigtaverner — APPROVED** [review]

**Pull request merged**